winpr/libwinpr/utils/wlog/Layout.c

```diff
@@ -57,201 +57,137 @@ static void WLog_PrintMessagePrefix(wLog* log, wLogMessage* message, const char*
 	va_end(args);
 }
 
+static const char* get_tid(void* arg)
+{
+	char* str = arg;
+	size_t tid;
+#if defined __linux__ && !defined ANDROID
+	/* On Linux we prefer to see the LWP id */
+	tid = (size_t)syscall(SYS_gettid);
+#else
+	tid = (size_t)GetCurrentThreadId();
+#endif
+	sprintf(str, "%08" PRIxz, tid);
+	return str;
+}
+
+static BOOL log_invalid_fmt(const char* what)
+{
+	fprintf(stderr, "Invalid format string '%s'\n", what);
+	return FALSE;
+}
+
+static BOOL check_and_log_format_size(char* format, size_t size, size_t index, size_t add)
+{
+	/* format string must be '\0' terminated, so abort at size - 1 */
+	if (index + add + 1 >= size)
+	{
+		fprintf(stderr,
+		        "Format string too long ['%s', max %" PRIuz ", used %" PRIuz ", adding %" PRIuz
+		        "]\n",
+		        format, size, index, add);
+		return FALSE;
+	}
+	return TRUE;
+}
+
+static BOOL check_and_log_format_args(size_t size, size_t index, size_t add)
+{
+	if (index + add > size)
+	{
+		fprintf(stderr,
+		        "Too many format string arguments [max %" PRIuz ", used %" PRIuz ", adding %" PRIuz
+		        "]\n",
+		        size, index, add);
+		return FALSE;
+	}
+	return TRUE;
+}
+
 BOOL WLog_Layout_GetMessagePrefix(wLog* log, wLogLayout* layout, wLogMessage* message)
 {
-	char* p;
-	int index;
-	int argc = 0;
-	void* args[32] = { 0 };
-	char format[256] = { 0 };
-	SYSTEMTIME localTime;
+	size_t index = 0;
+	unsigned argc = 0;
+	const void* args[32] = { 0 };
+	char format[WLOG_MAX_PREFIX_SIZE] = { 0 };
+	char tid[32] = { 0 };
+	SYSTEMTIME localTime = { 0 };
 
 	WINPR_ASSERT(layout);
 	WINPR_ASSERT(message);
 
 	GetLocalTime(&localTime);
-	index = 0;
-	p = (char*)layout->FormatString;
+	const char* p = layout->FormatString;
 
+	struct format_option
+	{
+		const char* fmt;
+		const char* replace;
+		const char* (*fkt)(void*);
+		const void* arg;
+	};
+
+	const struct format_option options[] = {
+		{ "%ctx", "%s", log->custom, log->context },                    /* log context */
```

Comment:
In this workflow, what about searching for `%` in `p`, and then trying to match options. Doing this we avoid scanning the beginning of the string. 
For example if prefix is "start of string %ml", then we will browse the _whole_ prefix numerous times (passing over `start of string ` with `strstr()`) until we match the `%ml` option which is the last one.
We may also rearrange the options so that the most commonly used are present first (if that's not already the case).

Reply:
@hardening ok, implemented `bsearch` for option comparison. this should reduce the number of required comparisons drastically.